### PR TITLE
Update tx pool ordering logic according to EIP-1559

### PR DESCRIPTION
**Pull request by begmaroman**
# Description

Update tx pool ordering logic according to EIP-1559 spec. 

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

File: types/encoding.go [hunk 1]
```diff
@@ -9,6 +9,10 @@ import (
 	"github.com/0xPolygon/polygon-edge/helper/hex"
 )
 
+// ParseUint64orHex parses the given string as uint64 in hex
+// It should go to the common package from the logical perspective
+// as well as avoiding cycle imports.
+// DEPRECATED. Use common.ParseUint64orHex.
 func ParseUint64orHex(val *string) (uint64, error) {
```
Comment (Stefan-Ethernal) [inline types/encoding.go:16]: I hope we are about to remove this function through some of the follow-up PRs.

File: helper/common/common.go [hunk 2]
```diff
@@ -79,6 +79,22 @@ func ConvertUnmarshalledUint(x interface{}) (uint64, error) {
 	}
 }
 
+func ParseUint64orHex(val *string) (uint64, error) {
```
Comment (Stefan-Ethernal) [inline helper/common/common.go:84]: Add a comment?

**Review by Stefan-Ethernal — APPROVED** [review]
Looks good.

Slightly off topic question, however EIP-1559 related one:
should we check if London hard-fork is enabled e.g. in `txpool.validateTx` and if it is not and we found `DynamicFeeTx` to return an error?
Also worth double checking is there any other place where we should explicitly check is London hard-fork enabled.

File: types/encoding.go (hunk 1 above)
Comment (begmaroman) [reply types/encoding.go:16]: Yeah, we will remove it

**Comment by begmaroman** [comment]
@Stefan-Ethernal I think it still should be possible to send legacy txs even if the hardfork is enabled.

**Comment by Stefan-Ethernal** [comment]
@begmaroman it is not a question about sending legacy transactions, but dynamic fee transactions if EIP-1559 hardfork is not enabled. My point is that we should have a validation in tx pool if EIP-1559 hardfork is not enabled. Makes sense?

**Comment by begmaroman** [comment]
@Stefan-Ethernal good point. Would you be able to add this comment to the parent PR since this is a little bit out of the current scope?

File: helper/common/common.go (hunk 2 above)
Comment (begmaroman) [reply helper/common/common.go:84]: Done

File: types/transaction.go [hunk 3]
```diff
@@ -150,7 +151,13 @@ func (t *Transaction) Cost() *big.Int {
 //   - or calculate a value with formula: min(gasFeeCap, gasTipCap * baseFee);
 func (t *Transaction) GetGasPrice(baseFee uint64) *big.Int {
 	if (t.GasPrice != nil && t.GasPrice.BitLen() > 0) || baseFee == 0 {
-		return new(big.Int).Set(t.GasPrice)
```
Comment (igorcrevar) [inline types/transaction.go:153]: Is this same as
```
if t.GasPrice != nil && t.GasPrice.BitLen() > 0 {
    return new(big.Int).Set(t.GasPrice)
} else if baseFee == 0  {
    return new(big.Int)
}
```
? somehow it is easier for me if there is one less if inside first if block

Comment (begmaroman) [reply types/transaction.go:153]: Done

File: helper/common/common.go [hunk 4]
```diff
@@ -79,6 +79,24 @@ func ConvertUnmarshalledUint(x interface{}) (uint64, error) {
 	}
 }
 
+// ParseUint64orHex parses the given uint64 hex string into the number.
+// It can parse the string with 0x prefix as well.
+func ParseUint64orHex(val *string) (uint64, error) {
```
Comment (goran-ethernal) [inline helper/common/common.go:84]: Don't we have the same function in `types` package?

**Review by goran-ethernal — APPROVED** [review]
Except for the minor comment LGTM

File: helper/common/common.go (hunk 4 above)
Comment (Stefan-Ethernal) [reply helper/common/common.go:84]: Yes, we do have it, but as Roman explained, the aim is to remove that one in some of the following PRs and use this one instead.
https://github.com/0xPolygon/polygon-edge/pull/1249/files#diff-c99d168ac05d1f4057606db9fa5ac73a6c7f5d00ef27273a912cb6cb174a8a45R12-R15

Comment (goran-ethernal) [reply helper/common/common.go:84]: Ah, OK, didn't see that he already explained that somewhere.

**Pull request merged**